framework/src/org/checkerframework/framework/util/ViewpointAdapter.java

```diff
@@ -0,0 +1,341 @@
+package org.checkerframework.framework.util;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.element.TypeParameterElement;
+import javax.lang.model.type.DeclaredType;
+import javax.lang.model.type.TypeKind;
+import org.checkerframework.dataflow.qual.SideEffectFree;
+import org.checkerframework.framework.type.AnnotatedTypeFactory;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedArrayType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedDeclaredType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedNullType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedTypeVariable;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedWildcardType;
+import org.checkerframework.framework.type.AnnotatedTypeReplacer;
+import org.checkerframework.javacutil.ErrorReporter;
+import org.checkerframework.javacutil.Pair;
+
+/**
+ * Utility class to perform viewpoint adaptation.
+ *
+ * @param <Modifier> Represents modifier on which viewpoint adaptation is performed, namely {@link
+ *     AnnotationMirror} on typechecking side and {@code Slot} on inference side
+ */
+public abstract class ViewpointAdapter<Modifier> {
+
+    // This prevents calling combineTypeWithType on type variable if it is an upper bound
+    // of another type variable. We only viewpoint adapt type variable that is not upper-bound.
+    protected boolean isTypeVarExtends = false;
+
+    /**
+     * Viewpoint adapt declared type to receiver type, and return the result atm
+     *
+     * @param receiver receiver type
+     * @param declared declared type
+     * @param f {@link AnnotatedTypeFactory} of concrete type system
+     * @return {@link AnnotatedTypeMirror} after viewpoint adaptation
+     */
+    public <TypeFactory extends AnnotatedTypeFactory> AnnotatedTypeMirror combineTypeWithType(
+            AnnotatedTypeMirror receiver, AnnotatedTypeMirror declared, TypeFactory f) {
```

Comment:
What is the best place to hold `AnnotationMirror`? Maybe pasing `AnnotatedTypeFactory` is not needed.